recipes/gdal/all/conanfile.py

```diff
@@ -298,13 +303,17 @@ def requirements(self):
         if self.options.get_safe("with_exr"):
             self.requires("openexr/2.5.2")
 
+    def _validate_dependency_graph(self):
+        if self.options.with_qhull and self.options["qhull"].reentrant:
+            raise ConanInvalidConfiguration("gdal depends on non-reentrant qhull.")
+
     def build_requirements(self):
-        if self.settings.os == "Windows" and self.settings.compiler != "Visual Studio" and \
-           "CONAN_BASH_PATH" not in os.environ and tools.os_info.detect_windows_subsystem() != "msys2":
+        if tools.os_info.is_windows and self.settings.compiler != "Visual Studio" and \
+           not tools.get_env("CONAN_BASH_PATH") and tools.os_info.detect_windows_subsystem() != "msys2":
```

Comment:
```suggestion
           not tools.get_env("CONAN_BASH_PATH"):
```
I would remove the `tools.os_info.detect_windows_subsystem` in the future.
There have been problems with it detecting the bash from git or anaconda.
Removing it and always installing msys2 is the safe bet, imho.

Reply:
What do you think of this?

```python
    def build_requirements(self):
        if tools.os_info.is_windows:
            if self.settings.compiler != "Visual Studio" and not tools.get_env("CONAN_BASH_PATH"):
                self.build_requires("msys2/20200517")
        else:
            self.build_requires("libtool/2.4.6")
            self.build_requires("pkgconf/1.7.3")
```

Reply:
Personally, I prefer the following. This puts the `msys2` requirement explicitly under `tools.os_info.is_windows`.
```python
    def build_requirements(self):
        if tools.os_info.is_windows:
            if self.settings.compiler != "Visual Studio" and not tools.get_env("CONAN_BASH_PATH"):
                self.build_requires("msys2/20200517")
        else:
            self.build_requires("libtool/2.4.6")
            self.build_requires("pkgconf/1.7.3")
```

Reply:
What happen if `tools.get_env("CONAN_BASH_PATH")` is True but pkgconf, libtool or autoconf is missing?

More robust test?

```python
    def build_requirements(self):
        if tools.os_info.is_windows and self.settings.compiler != "Visual Studio" and \
           not tools.get_env("CONAN_BASH_PATH"):
            self.build_requires("msys2/20200517")
        elif self.settings.compiler != "Visual Studio":
            self.build_requires("libtool/2.4.6")
            self.build_requires("pkgconf/1.7.3")
```

Reply:
Good point. libtool can be missing on linux too.
```python
    def build_requirements(self):
        self.build_requires("libtool/2.4.6")
        self.build_requires("pkgconf/1.7.3")
        if tools.os_info.is_windows and self.settings.compiler != "Visual Studio" and \
           not tools.get_env("CONAN_BASH_PATH"):
            self.build_requires("msys2/20200517")
```

Reply:
libtool and pkgconf are packaged with msys2 I think. I don't have any issue building gdal with mingw@windows, and msys2 only as build requirements.

Adding those two build requirements for MinGW would lead to build from source:
- ninja
- meson
- pkgconf
- m4
- autoconf
- automake
- libtool

and I faced some issues building some of them with MinGW if build_type=Debug

I prefer:

```python
    def build_requirements(self):
        if self.settings.compiler != "Visual Studio":
            if tools.os_info.is_windows and not tools.get_env("CONAN_BASH_PATH"):
                self.build_requires("msys2/20200517")
            else:
                self.build_requires("libtool/2.4.6")
                self.build_requires("pkgconf/1.7.3")
```

Reply:
Ok.
re your build problems, were these recent?
If you ever reproduce them again, please open an issue. :smile:

Reply:
Ok, tested 2 minutes ago with MinGW: m4 fails to link libm4.a into m4.exe in Release build type. I'll open an issue.

Reply:
Thanks, I guess :rofl:

Reply:
https://github.com/conan-io/conan-center-index/issues/2915 :smile: